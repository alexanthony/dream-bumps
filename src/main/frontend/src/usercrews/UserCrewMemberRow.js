// @flow
import React, {PropTypes, Component} from 'react';
import {connect} from 'react-redux';
import {TableRow, TableRowColumn} from 'material-ui/Table';
import RaisedButton from 'material-ui/RaisedButton';

import seatNumberToName from '../utils/seatNumberToName';
import {cancelBuyMember, doSellRower} from './actions';
import {getCrew, getCrewMemberName, getBuyMemberId, isMarketOpen} from '../rootReducer';
import {fetchCrewMembers, fetchCrew} from '../crews/actions';

class UserCrewMemberRow extends Component {
  componentDidMount() {
    if (this.props.member) {
      this.loadCrew(this.props);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps && nextProps.member && (!this.props.member || nextProps.member.crewId !== this.props.member.crewId)) {
      this.loadCrew(nextProps);
    }
  }

  loadCrew(props) {
    if (this.props.member.crewId) {
      this.props.fetchCrewMembers(this.props.member.crewId);
      this.props.fetchCrew(this.props.member.crewId);
    }
  }

  render() {
    const {
      member,
      crew,
      crewMemberName,
      onClickBuy,
      buyMemberId,
      cancelBuyMember,
      doSellRower,
      marketOpen,
      canBuySell
    } = this.props;
    return (
      <TableRow>
        <TableRowColumn>{seatNumberToName(member.seat)}</TableRowColumn>
        <TableRowColumn>
          {
            crew ?
              crew.name
            :
              '-'
          }
        </TableRowColumn>
        <TableRowColumn>{crewMemberName}</TableRowColumn>
        <TableRowColumn>
          {
            crew ?
              crew.price
            :
              ''
          }
        </TableRowColumn>
        {
          marketOpen && canBuySell ?
            <TableRowColumn>
              {
                member.crewId ?
                  <RaisedButton label="Sell" onClick={e => doSellRower(member.crewId, member.id)}/>
                :
                  member.id === buyMemberId ?
                    <RaisedButton label="Cancel" onClick={cancelBuyMember}/>
                  :
                    buyMemberId ?
                      null
                    :
                      <RaisedButton label="Buy" onClick={onClickBuy}/>
              }
            </TableRowColumn>
          :
            null
        }
      </TableRow>
    )
  }
}

// const CurrentUserCrewMemberRow = ({
//   member,
//   crew,
//   crewMemberName,
//   onClickBuy,
//   onClickCancel,
//   buyMemberId,
//   cancelBuyMember,
//   doSellRower,
//   marketOpen,
//   canBuySell
// }) => (
//   <TableRow>
//     <TableRowColumn>{seatNumberToName(member.seat)}</TableRowColumn>
//     <TableRowColumn>
//       {
//         crew ?
//           crew.name
//         :
//           '-'
//       }
//     </TableRowColumn>
//     {/*<TableRowColumn>{crewMemberName}</TableRowColumn>*/}
//     <TableRowColumn>
//       {
//         crew ?
//           crew.price
//         :
//           ''
//       }
//     </TableRowColumn>
//     {
//       marketOpen && canBuySell ?
//         <TableRowColumn>
//           {
//             member.crewId ?
//               <RaisedButton label="Sell" onClick={e => doSellRower(member.crewId, member.id)}/>
//             :
//               member.id === buyMemberId ?
//                 <RaisedButton label="Cancel" onClick={cancelBuyMember}/>
//               :
//                 buyMemberId ?
//                   null
//                 :
//                   <RaisedButton label="Buy" onClick={onClickBuy}/>
//           }
//         </TableRowColumn>
//       :
//         null
//     }
//   </TableRow>
// );

UserCrewMemberRow.propTypes = {
  member: PropTypes.object,
  crew: PropTypes.object,
  crewMemberName: PropTypes.string,
  onClickBuy: PropTypes.func,
  buyMemberId: PropTypes.number,
  cancelBuyMember: PropTypes.func,
  doSellRower: PropTypes.func,
  marketOpen: PropTypes.bool,
  canBuySell: PropTypes.bool,
  fetchCrewMembers: PropTypes.func,
  fetchCrew: PropTypes.func
};

const mapStateToProps = (state, {member}) => ({
  crew: getCrew(state)(member.crewId),
  crewMemberName: getCrewMemberName(state)(member.crewId, member.seat) || '',
  buyMemberId: getBuyMemberId(state),
  marketOpen: isMarketOpen(state)
});

export default connect(mapStateToProps, {cancelBuyMember, doSellRower, fetchCrewMembers, fetchCrew})(UserCrewMemberRow);
