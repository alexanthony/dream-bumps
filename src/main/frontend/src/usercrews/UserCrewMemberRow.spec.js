import React from 'react';
import { shallow } from 'enzyme';
import { UserCrewMemberRow } from './UserCrewMemberRow';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const cancelBuyMember = jest.fn();
  const fetchCrewMembers = jest.fn();
  const onClickBuy = jest.fn();
  const doSellRower = jest.fn();
  const fetchCrew = jest.fn();
  const member = {id: 5, seat: 9, crewId: 8};
  const crew = {name: 'Crew Name', id: 8, price: 123.56};
  const component = shallow(
    <UserCrewMemberRow
      member={member}
      crew={crew}
      crewMemberName="Something"
      onClickBuy={onClickBuy}
      buyMemberId={5}
      cancelBuyMember={cancelBuyMember}
      doSellRower={doSellRower}
      marketOpen
      canBuySell
      fetchCrewMembers={fetchCrewMembers}
      fetchCrew={fetchCrew}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});

test('Renders properly', () => {
  const cancelBuyMember = jest.fn();
  const fetchCrewMembers = jest.fn();
  const onClickBuy = jest.fn();
  const doSellRower = jest.fn();
  const fetchCrew = jest.fn();
  const member = {id: 5, seat: 9, crewId: 8};
  const crew = {name: 'Crew Name', id: 8, price: 123.56};
  const component = shallow(
    <UserCrewMemberRow
      member={member}
      crew={crew}
      crewMemberName="Something"
      onClickBuy={onClickBuy}
      cancelBuyMember={cancelBuyMember}
      doSellRower={doSellRower}
      marketOpen={false}
      canBuySell={false}
      fetchCrewMembers={fetchCrewMembers}
      fetchCrew={fetchCrew}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});

test('Renders properly', () => {
  const cancelBuyMember = jest.fn();
  const fetchCrewMembers = jest.fn();
  const onClickBuy = jest.fn();
  const doSellRower = jest.fn();
  const fetchCrew = jest.fn();
  const member = {id: 5, seat: 9, crewId: 8};
  const crew = {name: 'Crew Name', id: 8, price: 123.56};
  const component = shallow(
    <UserCrewMemberRow
      member={member}
      crew={crew}
      crewMemberName="Something"
      onClickBuy={onClickBuy}
      cancelBuyMember={cancelBuyMember}
      doSellRower={doSellRower}
      marketOpen
      canBuySell
      fetchCrewMembers={fetchCrewMembers}
      fetchCrew={fetchCrew}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});
