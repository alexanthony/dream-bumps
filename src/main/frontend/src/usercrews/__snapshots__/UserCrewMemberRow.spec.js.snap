exports[`test Renders properly 1`] = `
<TableRow
  displayBorder={true}
  hoverable={false}
  hovered={false}
  selectable={true}
  selected={false}
  striped={false}>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
        "width": "45px",
      }
    }>
    Cox
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
      }
    }>
    Crew Name
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
      }
    }>
    Something
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
        "textAlign": "right",
        "width": "75px",
      }
    }>
    123.56
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
        "width": "75px",
      }
    }>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Cancel"
      labelPosition="after"
      labelStyle={
        Object {
          "paddingLeft": "2px",
          "paddingRight": 0,
        }
      }
      onClick={[Function]}
      primary={true}
      secondary={false}
      style={
        Object {
          "minWidth": "70px",
        }
      } />
  </TableRowColumn>
</TableRow>
`;

exports[`test Renders properly 2`] = `
<TableRow
  displayBorder={true}
  hoverable={false}
  hovered={false}
  selectable={true}
  selected={false}
  striped={false}>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
        "width": "45px",
      }
    }>
    Cox
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
      }
    }>
    Crew Name
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
      }
    }>
    Something
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
        "textAlign": "right",
        "width": "75px",
      }
    }>
    123.56
  </TableRowColumn>
</TableRow>
`;

exports[`test Renders properly 3`] = `
<TableRow
  displayBorder={true}
  hoverable={false}
  hovered={false}
  selectable={true}
  selected={false}
  striped={false}>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
        "width": "45px",
      }
    }>
    Cox
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
      }
    }>
    Crew Name
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
      }
    }>
    Something
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": "5px",
        "textAlign": "right",
        "width": "75px",
      }
    }>
    123.56
  </TableRowColumn>
  <TableRowColumn
    hoverable={false}
    style={
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
        "width": "75px",
      }
    }>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Sell"
      labelPosition="after"
      onClick={[Function]}
      primary={true}
      secondary={false}
      style={
        Object {
          "minWidth": "70px",
        }
      } />
  </TableRowColumn>
</TableRow>
`;
